allennlp/training/deepspeed/trainer.py

```diff
@@ -0,0 +1,703 @@
+        if you wish to load a model for inference/load parts of a model into a new
+        computation graph, you should use the native Pytorch functions:
+        ` model.load_state_dict(torch.load("/path/to/model/weights.th"))`
+        If `self._serialization_dir` does not exist or does not contain any checkpointed weights,
+        this function will do nothing and return 0.
+        # Returns
+        epoch: `int`
+            The epoch at which to resume training, which should be one after the epoch
+            in the saved training state.
+        """
+        checkpoint_id, model_state, training_state = self._checkpointer.restore_checkpoint()
+
+        if not training_state:
+            # No checkpoint to restore, start at 0
+            return 0
+
+        self.model.load_state_dict(model_state)
+        self.model_engine.load_checkpoint(self._serialization_dir, checkpoint_id)
+        # self.optimizer.load_state_dict(training_state["optimizer"])
+        # training_util.move_optimizer_to_cuda(self.optimizer)
+
+        # Currently the `training_state` contains a serialized `MetricTracker`.
+        if "metric_tracker" in training_state:
+            self._metric_tracker.load_state_dict(training_state["metric_tracker"])
+        # It used to be the case that we tracked `val_metric_per_epoch`.
+        elif "val_metric_per_epoch" in training_state:
+            self._metric_tracker.clear()
+            self._metric_tracker.add_metrics(training_state["val_metric_per_epoch"])
+        # And before that we didn't track anything.
+        else:
+            self._metric_tracker.clear()
+
+        if isinstance(training_state["epoch"], int):
+            epoch_to_return = training_state["epoch"] + 1
+        else:
+            epoch_to_return = int(training_state["epoch"].split(".")[0]) + 1
+
+        # For older checkpoints with batch_num_total missing, default to old behavior where
+        # it is unchanged.
+        batch_num_total = training_state.get("batch_num_total")
+        if batch_num_total is not None:
+            self._batch_num_total = batch_num_total
+
+        return epoch_to_return
+
+    @classmethod
+    def from_partial_objects(
+        cls,
+        model: Model,
+        serialization_dir: str,
+        data_loader: DataLoader,
+        deepspeed_config: DeepspeedConfig,
+
+        validation_data_loader: DataLoader = None,
+        local_rank: int = 0,
+        patience: int = None,
+        validation_metric: str = "-loss",
+        num_epochs: int = 20,
+        cuda_device: Optional[Union[int, torch.device]] = None,
+        distributed: bool = None,
+        world_size: int = 1,
+        num_gradient_accumulation_steps: int = 1,
+        no_grad: List[str] = None,
+        optimizer: Lazy[Optimizer] = None,
+        deepspeed_optimizer: Dict[str, Any] = None,
+        deepspeed_args: Lazy[DeepspeedArgs] = Lazy(DeepspeedArgs),
```

Comment:
Do these have to be lazy?

Reply:
Yeah, they need the batch size and number of gradient accumulation steps (which I get from this constructor and the dataloader), so Lazy felt right here. I could set the attributes manually if we're specifically avoiding Lazy objects.

Reply:
We don't go out of our way to avoid using them, but if we don't have to use them, we don't, because it's hard to debug what's going on. I'll leave it to your judgement. I thought these were just user-specified data classes, in which case it would be a bit weird to make them lazy, but if they are filled automatically, it makes more sense.